py/server/deephaven_internal/plugin/object/__init__.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+
+import jpy
+
+from typing import Optional, List, Any
+from deephaven.plugin.object_type import Exporter, ObjectType, Reference, MessageStream, FetchOnlyObjectType
+from deephaven._wrapper import JObjectWrapper, wrap_j_object
+
+JReference = jpy.get_type('io.deephaven.plugin.type.Exporter$Reference')
+JExporterAdapter = jpy.get_type('io.deephaven.server.plugin.python.ExporterAdapter')
+JMessageStream = jpy.get_type('io.deephaven.plugin.type.ObjectType$MessageStream')
+
+
+def _adapt_reference(ref: JReference) -> Reference:
+    return Reference(ref.index(), ref.type().orElse(None))
+
+
+def _unwrap(object):
+    # todo: we should have generic unwrapping code ABC
+    if isinstance(object, JObjectWrapper):
+        return object.j_object
+    return object
+
+
+class ExporterAdapter(Exporter):
+    """Python implementation of Exporter that delegates to its Java counterpart."""
+
+    def __init__(self, exporter: JExporterAdapter):
+        self._exporter = exporter
+
+    def reference(self, obj: Any, allow_unknown_type: bool = True, force_new: bool = True) -> Optional[Reference]:
+        obj = _unwrap(obj)
+        if isinstance(obj, jpy.JType):
+            ref = self._exporter.reference(obj, allow_unknown_type, force_new)
+        else:
+            ref = self._exporter.referencePyObject(obj, allow_unknown_type, force_new)
+        return _adapt_reference(ref) if ref else None
+
+    def __str__(self):
+        return str(self._exporter)
+
+
+class ClientResponseStreamAdapter(MessageStream):
+    """Python implementation of MessageStream that delegates to its Java counterpart"""
+    def __init__(self, wrapped: JMessageStream):
+        self._wrapped = wrapped
+
+    def on_data(self, payload: bytes, references: List[Any]) -> None:
+        self._wrapped.onData(payload, [_unwrap(ref) for ref in references])
+
+    def on_close(self) -> None:
+        self._wrapped.onClose()
+
+
+class ServerRequestStreamAdapter(MessageStream):
+    """Wraps Python server MessageStream implementations to correctly adapt objects coming from the client
+    """
+    def __init__(self, wrapped: MessageStream):
+        self._wrapped = wrapped
+
+    def on_data(self, payload:bytes, references: List[Any]) -> None:
+        self._wrapped.on_data(payload, [wrap_j_object(ref) for ref in references])
+
+    def on_close(self) -> None:
+        self._wrapped.on_close()
+
+
+class ObjectTypeAdapter:
+    """Python type that Java's ObjectTypeAdapter will call in order to communicate with a Python ObjectType instance."""
+
+    def __init__(self, user_object_type: ObjectType):
+        self._user_object_type = user_object_type
+
+    def is_type(self, obj) -> bool:
+        return self._user_object_type.is_type(obj)
+
+    def is_fetch_only(self) -> bool:
+        return isinstance(self._user_object_type, FetchOnlyObjectType)
```

Comment:
I don't know how important it is; but arguably, we could also check if it has a to_bytes method and it might work against existing plugins.

Reply:
The fact that we also moved Callback, and also took away the default False values for the exporter methods makes me suspect that most existing plugins will need an update anyway.